### DISCUSSION TITLE
sshd_lineinfile - oval.template - issue

### DISCUSSION THREAD

**Discussion by rumch-se** [ComplianceAsCode/content#10704] [General]
Hello @yuumasato  

I found that you are creator of the oval template in sshd_lineinfile template, and for it you are using sshd_oval_check. I think that this template does not work as expected.  You are re-suing the macro sshd_oval_check, created by @matejak 
May you check this (for example you can use as a reference the rule  sshd_set_keepalive_0)?
What is the problem:
When we have in the file /etc/ssh/sshd_config a row for example 
\# ClientAliveCountMax 3
the oval check will pass, but this is not correct because in front of the parameter we have a symbol which will be interpreted as comments. Please check this (https://linux.die.net/man/5/sshd_config). The oval check should fail.
The remediation part after should remove the symbol for comment or to add a line to a sshd_config with correct settings.
I think that all rules which use this template will be / are impacted. 

Have a nice day
Rumen

**Comment by ggbecker** [comment]
In this case you need to provide the report with the OVAL results. It could a different condition in the OVAL that is making the rule to pass, for example:

```
   <criteria comment="sshd is configured correctly or is not installed" operator="OR">
     <criteria comment="sshd is not installed" operator="AND">
        <extend_definition comment="sshd is not required or requirement is unset"
          definition_ref="sshd_not_required_or_unset" />
        <extend_definition comment="rpm package openssh-server removed"
          definition_ref="package_openssh-server_removed" />
     </criteria>
     <criteria comment="sshd is installed and configured" operator="AND">
        <extend_definition comment="sshd is required or requirement is unset"
          definition_ref="sshd_required_or_unset" />
        <extend_definition comment="rpm package openssh-server installed"
          definition_ref="package_openssh-server_installed" />
        <criteria comment="sshd is configured correctly" operator="AND">
          <criteria comment="the configuration is correct if it exists" operator="AND"><criterion comment="Check the ClientAliveCountMax in /etc/ssh/sshd_configif any"
            test_ref="test_sshd_set_keepalive_0" /><criterion comment="Check the ClientAliveCountMax in /etc/ssh/sshd_config.dif any"
            test_ref="test_sshd_set_keepalive_0_config_dir" />
            
          </criteria>
          <criterion comment="the configuraton exists" test_ref="test_ClientAliveCountMax_present_sshd_set_keepalive_0" />
          
        </criteria>
      </criteria>
    </criteria>
```

**Comment by rumch-se** [comment]
Hello @ggbecker , @yuumasato and @matejak 
This is a screenshot which presents the issue. Because I am facing this problem for a 1st time, how can I generate this report?
I have executed the command
oscap --verbose INFO xccdf eval .... and report is attached 
Have a nice day
Rumen
![template_1](https://github.com/ComplianceAsCode/content/assets/77793453/36ce6a98-25c9-4e8c-bf7c-847628bd18b0)
[report.txt](https://github.com/ComplianceAsCode/content/files/11706270/report.txt)

**Reply by ggbecker** [reply]
add `--report report.html --oval-results` to your command line. In the report.html there should be the OVAL results.

Consider also exporting the `--results-arf results-arf.xml` file.

**Comment by rumch-se** [comment]
Hello @ggbecker, @yuumasato and @matejak 
Here are the generated reports.
Have a nice day
 
[reports.zip](https://github.com/ComplianceAsCode/content/files/11706386/reports.zip)

**Comment by ggbecker** [comment]
My initial guess is that this piece of jinja condition:

https://github.com/ComplianceAsCode/content/blob/fa777d42fe9ac21eda9595fc8a6bb72c6b0c73b9/shared/macros/10-oval.jinja#L611-L617

is missing in:

https://github.com/ComplianceAsCode/content/blob/fa777d42fe9ac21eda9595fc8a6bb72c6b0c73b9/shared/macros/10-oval.jinja#L929-L934

here is the report generated using the [oscap-report](https://github.com/OpenSCAP/openscap-report) tool and the ARF file you provided.

(rename it to report.html)
[report.txt](https://github.com/ComplianceAsCode/content/files/11706472/report.txt)

**Comment by rumch-se** [comment]
Hello @ggbecker 
There is another rule which does not use template but it includes ansible, oval, bash, etc. The name of the rule is sshd_set_keepalive. You can see that his oval file section includes a similar check of the product. Again in this case there is a issue. Screen shot is provided.   

And my concern is this regexp in the oval
<ind:pattern operation="pattern match">^[\s]*(?i)ClientAliveCountMax[\s]+([\d]+)[\s]*(?:#.*)?$</ind:pattern> 
![template_2](https://github.com/ComplianceAsCode/content/assets/77793453/c3fff7c6-ea16-4f5c-8fe5-104725a29c7d)

**Reply by rumch-se** [reply]
@ggbecker 
- the regexp is looking for the existence of the ClientAliveCountMax in the file, but it does not care what kind of characters/values we have before and after the ClientAliveCountMax

**Reply by ggbecker** [reply]
In this case the regex should be: `^[ \t]*(?i)ClientAliveCountMax(?-i)[ \t]+(.+?)[ \t]*(?:$|#)` (from what's derived here: https://github.com/ComplianceAsCode/content/blob/fa777d42fe9ac21eda9595fc8a6bb72c6b0c73b9/shared/macros/10-oval.jinja#L971)

as you can see here: 
https://regex101.com/r/PBZApj/1

**Reply by ggbecker** [reply]
I believe that every regex in the following list needs to be updated to be similar to the example above:

I search for `^[\s]*(?i)` in `shared.xml` files in the project:
```
linux_os/guide/services/ssh/ssh_server/sshd_disable_rhosts_rsa/oval/shared.xml:
  43      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  44:     <ind:pattern operation="pattern match">^[\s]*(?i)RhostsRSAAuthentication(?-i)[\s]+no[\s]*(?:#.*)?$</ind:pattern>
  45      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_set_idle_timeout/oval/shared.xml:
  51      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  52:     <ind:pattern operation="pattern match">^[\s]*(?i)ClientAliveInterval[\s]+(\d+)[\s]*(?:#.*)?$</ind:pattern>
  53      <ind:instance datatype="int">1</ind:instance>

  66      <ind:filename operation="pattern match">.*\.conf$</ind:filename>
  67:     <ind:pattern operation="pattern match">^[\s]*(?i)ClientAliveInterval[\s]+(\d+)[\s]*(?:#.*)?$</ind:pattern>
  68      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_set_login_grace_time/oval/shared.xml:
  42      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  43:     <ind:pattern operation="pattern match">^[\s]*(?i)LoginGraceTime[\s]+(\d+)[\s]*(?:#.*)?$</ind:pattern>
  44      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_set_max_auth_tries/oval/shared.xml:
  31      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  32:     <ind:pattern operation="pattern match">^[\s]*(?i)MaxAuthTries[\s]+(\d+)[\s]*(?:#.*)?$</ind:pattern>
  33      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_set_max_sessions/oval/shared.xml:
  42      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  43:     <ind:pattern operation="pattern match">^[\s]*(?i)MaxSessions[\s]+(\d+)[\s]*(?:#.*)?$</ind:pattern>
  44      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_use_approved_ciphers/oval/shared.xml:
  42      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  43:     <ind:pattern operation="pattern match">^[\s]*(?i)Ciphers(?-i)[\s]+([\w,-@]+)+[\s]*(?:#.*)?$</ind:pattern>
  44      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_use_approved_ciphers_ordered_stig/oval/shared.xml:
  33      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  34:     <ind:pattern operation="pattern match">^[\s]*(?i)Ciphers(?-i)[\s]+(?=[\w]+)(aes256-ctr(?=[\w,]+|$),?)?(aes192-ctr(?=[\w,]+|$),?)?(aes128-ctr)?[\s]*(?:#.*)?$</ind:pattern>
  35      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_use_approved_macs/oval/shared.xml:
  52      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  53:     <ind:pattern operation="pattern match">^[\s]*(?i)MACs(?-i)[\s]+([\w,-@]+)+[\s]*(?:#.*)?$</ind:pattern>
  54      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_use_approved_macs_ordered_stig/oval/shared.xml:
  33      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  34:     <ind:pattern operation="pattern match">^[\s]*(?i)MACs(?-i)[\s]+(?=[\w]+)(hmac-sha2-512(?=[\w,]+|$),?)?(hmac-sha2-256)?[\s]*(?:#.*)?$</ind:pattern>
  35      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/services/ssh/ssh_server/sshd_use_strong_kex/oval/shared.xml:
  49      <ind:filepath>/etc/ssh/sshd_config</ind:filepath>
  50:     <ind:pattern operation="pattern match">^[\s]*(?i)KexAlgorithms(?-i)[\s]+([\w,-@]+)+[\s]*(?:#.*)?$</ind:pattern>
  51      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/system/accounts/accounts-session/accounts_have_homedir_login_defs/oval/shared.xml:
  16      <ind:filepath>/etc/login.defs</ind:filepath>
  17:     <ind:pattern operation="pattern match">^[\s]*(?i)CREATE_HOME(?-i)[\s]+yes[\s]*(?:#.*)?$</ind:pattern>
  18      <ind:instance datatype="int">1</ind:instance>

linux_os/guide/system/accounts/accounts-session/accounts_logon_fail_delay/oval/shared.xml:
  16      <ind:filepath>/etc/login.defs</ind:filepath>
  17:     <ind:pattern operation="pattern match">^[\s]*(?i)FAIL_DELAY(?-i)[\s]+([^#\s]*)</ind:pattern>
  18      <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>

linux_os/guide/system/accounts/accounts-session/user_umask/accounts_umask_etc_csh_cshrc/oval/shared.xml:
  13      <ind:filepath>/etc/csh.cshrc</ind:filepath>
  14:     <ind:pattern operation="pattern match">^[\s]*(?i)UMASK(?-i)[\s]+([^#\s]*)</ind:pattern>
  15      <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>

linux_os/guide/system/permissions/restrictions/daemon_umask/umask_for_daemons/oval/shared.xml:
  13      <ind:filepath>/etc/init.d/functions</ind:filepath>
  14:     <ind:pattern operation="pattern match">^[\s]*(?i)UMASK(?-i)[\s]+([^#\s]*)</ind:pattern>
  15      <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>
```

Can you work on those items and update the regex accordingly? Note that most of these rules don't have test scenarios and some of them is not possible to use the template sshd_lineinfile because they require XCCDF variables to be used and the template does not support it right now.

And the OVAL for them, don't have the jinja if for sle12,sle15 products as I mentioned in previous comments.

**Reply by rumch-se** [reply]
Hello @ggbecker
No it did not work as you expected. 
I think we have to be more precise because we define a variable which we use it in case of remediation. In other words if the variable has a value = 3, the oval check has to pass only when we have
spaces  ClientAliveCountMax spaces 3 as a row in the /etc/ssh/sshd_config 
by default after the installation of sshd  in the configuration file we have 
\#ClientAliveCountMax 3 
but this is not valid. The regular expression is looking for this parameter, but it does not check if the parameter has a comment
The new picture is attached.  
Have a nice day
Rumen 
![template_3](https://github.com/ComplianceAsCode/content/assets/77793453/f29d5053-1ec6-4815-a973-ffe279634921)

**Reply by ggbecker** [reply]
Can you check if there is no other configuration file under `/etc/ssh/sshd_config.d/` that could contain the same configuration set?

You really need to learn how to use the reports and the results to understand exactly which OVAL criteria is failing and which is not. If you look at the report.html generated by the oscap command using --oval-results, you should be able to inspect the exact details of what OVAL criteria has passed and what not.

**Reply by ggbecker** [reply]
This should fix the problem for at least this rule: https://github.com/ComplianceAsCode/content/pull/10706

**Reply by rumch-se** [reply]
Hello @ggbecker 
On my test machine I don't have any file in /etc/ssh/sshd_config.d/
here is ls 
sles-15-sp2:/etc/ssh # ls -la
total 620
drwxr-xr-x   2 root root   4096 Jun  9 12:09 .
drwxr-xr-x 117 root root  12288 May 17 08:28 ..
-rw-------   1 root root 577388 Jun  6  2020 moduli
-rw-r--r--   1 root root   2356 Jun  6  2020 ssh_config
-rw-------   1 root root   1381 Mar 16 09:30 ssh_host_dsa_key
-rw-r--r--   1 root root    604 Mar 16 09:30 ssh_host_dsa_key.pub
-rw-------   1 root root    505 Mar 16 09:30 ssh_host_ecdsa_key
-rw-r--r--   1 root root    176 Mar 16 09:30 ssh_host_ecdsa_key.pub
-rw-------   1 root root    411 Mar 16 09:30 ssh_host_ed25519_key
-rw-r--r--   1 root root     96 Mar 16 09:30 ssh_host_ed25519_key.pub
-rw-------   1 root root   2602 Mar 16 09:30 ssh_host_rsa_key
-rw-r--r--   1 root root    568 Mar 16 09:30 ssh_host_rsa_key.pub
-rw-r-----   1 root root   3909 Jun  9 12:13 sshd_config
sles-15-sp2:/etc/ssh #

**Reply by ggbecker** [reply]
in this case it might be the lack of sle15 in this line: https://github.com/ComplianceAsCode/content/pull/10706/files#diff-8134f071ec75dfb7c90732ea670fd6786c886445158db93fd330ac26ca7a6b39R9

**Reply by ggbecker** [reply]
I hope I gave you enough guidance here. There seems to have some inconsistencies across different files that need to be fixed for SLES products.

**Comment by rumch-se** [comment]
Hello @ggbecker 
With this PR https://github.com/ComplianceAsCode/content/pull/10353/files#diff-bf0e6efb88d44f16036a42436119df578d5190733c6ac8711ebd7516fd58a2bc  I was trying to be more precise in oval check
Have a nice day
Rumen
